### PR TITLE
Add applications api

### PR DESCRIPTION
**Pull request by zognin**
Handlers
- POST
  - Application (roleId, userId).
  - Does not check if roleId is a valid foreign key, left to prisma to handle.
- GET
  - List of applications and its relevant data. Refer to the types for more info.
  - Sorts the list by most taskNotifications, followed by latest date of latest stage.

**Note: Please run `yarn prisma generate` to regenerate the client, otherwise filtered relations count cannot compile**
References: 
- https://www.prisma.io/docs/concepts/components/prisma-client/aggregation-grouping-summarizing#filter-the-relation-count
- https://www.prisma.io/docs/concepts/components/preview-features/client-preview-features#enabling-a-prisma-client-preview-feature

Other notes
- Will do search in a different PR. May want to reuse some of the roles search logic.

### CODE REVIEW

File: utils/constants.ts [hunk 1]
```diff
@@ -1 +1,2 @@
 export const MIN_ROLE_YEAR = 1970;
+export const MIN_DATE = new Date(-8640000000000000);
```
Comment (sivayogasubramanian) [inline utils/constants.ts:4]: Let's add a comment here.

File: pages/api/applications/index.ts [hunk 2]
```diff
@@ -1,5 +1,174 @@
...
+  }
+
+  if (req.body.roleId === null || !Number.isInteger(req.body.roleId)) {
+    return MessageType.INVALID_ROLE_ID;
+  }
+
+  return null;
+}
+
+export default withAuthUser(handler);
```
Comment (sivayogasubramanian) [inline pages/api/applications/index.ts]: Add prisma error handling

File: types/application.ts [hunk 3]
```diff
@@ -2,10 +2,29 @@ import { ApplicationStageApplicationData } from './applicationStage';
...
 };
+
+export type ApplicationRoleData = {
+  id: number;
+  role: RoleData & { company: CompanyData };
+};
+
+export type ApplicationPostData = {
+  roleId: number;
+  userId: string;
```
Comment (sivayogasubramanian) [inline types/application.ts]: Why is there a userId here?

File: pages/api/applications/index.ts [hunk 4]
```diff
@@ -1,5 +1,174 @@
...
+    .status(HttpStatus.CREATED)
+    .json(createJsonResponse(newApplication, messages[MessageType.APPLICATION_CREATED_SUCCESSFULLY]));
+}
+
+function validatePostRequest(req: NextApiRequest): Nullable<MessageType> {
+  if (req.body.roleId === undefined) {
+    return MessageType.MISSING_ROLE_ID;
+  }
+
+  if (req.body.roleId === null || !Number.isInteger(req.body.roleId)) {
```
Comment (sivayogasubramanian) [inline pages/api/applications/index.ts:167]: Can you help to change all instances to `isInteger` to this?

**Review by sivayogasubramanian — APPROVED** [review]
LGTM! Pre-approving. Merge when the comments are resolved.

File: types/application.ts (hunk 3 above)
Comment (zognin) [reply types/application.ts]: Creation of application needs a userId, but I can also choose not to put it in here. Edit: I think I'll leave it there for clarity.

Comment (sivayogasubramanian) [reply types/application.ts]: Isn't the userId available in the handler?

Comment (sivayogasubramanian) [reply types/application.ts]: I feel like this should represent the shape of the JSON that we send over and not what prisma needs.

Comment (zognin) [reply types/application.ts]: Okay

**Pull request merged**